File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        for row in matrix:
            prefix = 0
            for col in range(len(row)):
                prefix += row[col]
                row[col] = prefix
        self.matrix = matrix
    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        aggr = 0
        for i in range(row1, row2+1):
            if col1:
                aggr += self.matrix[i][col2] - self.matrix[i][col1-1]
            else:
                aggr += self.matrix[i][col2]
        return aggr


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.cpp
class NumMatrix {
public:
    int grid[205][205]{};
    NumMatrix(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        
        for (int i = 1; i <= n; i++)
            for (int j = 1; j <= m; j++)
                grid[i][j] = grid[i - 1][j] + grid[i][j - 1] - grid[i - 1][j - 1] + matrix[i - 1][j - 1];
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return grid[row2 + 1][col2 + 1] - grid[row2 + 1][col1] - grid[row1][col2 + 1] + grid[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */